src/engine/tools.cpp

```diff
@@ -255,19 +255,32 @@ namespace fheroes2
     {
         const int dx = pt2.x - pt1.x;
         const int dy = pt2.y - pt1.y;
-        const uint32_t dist = static_cast<uint32_t>( std::hypot( std::abs( dx ), std::abs( dy ) ) );
-        // round up the integer division
-        const uint32_t length = ( step > 0 && dist >= step / 2 ) ? ( dist + step / 2 ) / step : 1;
-        const double moveX = dx / static_cast<double>( length );
-        const double moveY = dy / static_cast<double>( length );
+        const uint32_t dist = static_cast<uint32_t>( std::hypot( dx, dy ) );
+        // Round up the integer division and avoid the division by zero in calculation of total line points.
+        const uint32_t length = ( step > 0 ) ? ( dist + step / 2 ) / step : 0;
 
         std::vector<Point> line;
-        line.reserve( length );
 
-        for ( uint32_t i = 0; i <= length; ++i ) {
-            line.emplace_back( static_cast<int>( pt1.x + i * moveX ), static_cast<int>( pt1.y + i * moveY ) );
+        if ( length == 0 ) {
+            // If the length is equal to 0 than 'pt2' could be closer to 'pt1' than 'step'.
```

Comment:
I can suggest to slightly optimize the function: if the length is 1, we also don't need to do the rest calculations as we already have the needed points: 'pt1' and 'pt2'.
```suggestion
        if ( length <= 1 ) {
            // If the length is equal to 1 than we have only two points: 'pt1' and 'pt2'.
            // If the length is equal to 0 than 'pt2' could be closer to 'pt1' than 'step'.
```